### PR TITLE
optimize: add e2etests buildtags for e2e

### PR DESCRIPTION
**Pull request by 1046102779**
```shell
=== RUN   TestAuthentication
=== RUN   TestAuthentication/Valid_jwt_and_valid_seed
err: nats: no servers available for connection
    authentication_test.go:56: Did not expect error during connect nats: no servers available for connection
=== RUN   TestAuthentication/Valid_jwt_and_invalid_seed
err: dial tcp [::1]:4222: connect: network is unreachable
=== RUN   TestAuthentication/Invalid_jwt_and_valid_seed
err: nats: no servers available for connection
--- FAIL: TestAuthentication (0.00s)
    --- FAIL: TestAuthentication/Valid_jwt_and_valid_seed (0.00s)
    --- PASS: TestAuthentication/Valid_jwt_and_invalid_seed (0.00s)
    --- PASS: TestAuthentication/Invalid_jwt_and_valid_seed (0.00s)
FAIL
FAIL    github.com/dapr/components-contrib/tests/e2e/pubsub/jetstream/authentication    0.009s
FAIL
```

### CODE REVIEW

File: tests/e2e/pubsub/jetstream/authentication/authentication_test.go [hunk 1]
```diff
@@ -53,6 +54,11 @@ func TestAuthentication(t *testing.T) {
 			}
 			err := js.Init(md)
 			if err != nil && !tC.expectError {
+				if strings.Contains(err.Error(), "network is unreachable") ||
+					strings.Contains(err.Error(), "no servers available for connection") {
```
Comment (daixiang0) [inline tests/e2e/pubsub/jetstream/authentication/authentication_test.go:58]: Why it can not start as expected?

Comment (1046102779) [reply tests/e2e/pubsub/jetstream/authentication/authentication_test.go:58]: When using `go test -v ./...`, jetstream is not actually started locally. At present, if this unit test is to pass, the jetstream server must be started locally through ` docker-compose.yaml`.

Comment (1046102779) [reply tests/e2e/pubsub/jetstream/authentication/authentication_test.go:58]: Another problem:

I passed the unit tests last week, but after merging the community code yesterday, I found that the unit tests could not run, and there were the following errors:

image

image

**Comment by ItalyPaleAle** [comment]
Normally, tests in the `e2e` folder don't run when running `make test`.

However, I understand the issue and I think it should be fixed. But, if the server isn't ready, the test should be skipped entirely and not just avoiding a fatal log.

See how `binding.ipfs` does it: if you want to run the test, you need to pass an environmental variable; otherwise, the entire test is skipped:

https://github.com/dapr/components-contrib/blob/master/bindings/ipfs/ipfs_test.go#L36-L43

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Please change to use an env var to make the tests run. See example in `bindings.ipfs`: https://github.com/dapr/components-contrib/blob/master/bindings/ipfs/ipfs_test.go#L36-L43

**Comment by berndverst** [comment]
If this is an integration test I think we should use build tags! Please don't use environment variables to trigger the test unless this is needed for configuring the server etc, still I think build tags are the way to go.

FYI @ItalyPaleAle

**Comment by 1046102779** [comment]
[berndverst](https://github.com/berndverst)
[ItalyPaleAle](https://github.com/ItalyPaleAle)

/cc @berndverst @ItalyPaleAle. Thanks, i added a e2etest buildtag for the unit test jetstream  file.

**Comment by 1046102779** [comment]
> Please change to use an env var to make the tests run. See example in `bindings.ipfs`: https://github.com/dapr/components-contrib/blob/master/bindings/ipfs/ipfs_test.go#L36-L43

IPFS_TEST: "1"

```golang
      if utils.IsTruthy(os.Getenv("IPFS_TEST")) {
                  log.Println("IPFS_TEST env var is not set to a truthy value; skipping tests")
                  os.Exit(0)
      }
```

`IPFS_TEST=1` don't execute this unit test. Why not 0?

**Comment by ItalyPaleAle** [comment]
> IPFS_TEST=1 don't execute this unit test. Why not 0?

It should! Did you `export IPFS_TEST=1`?

> If this is an integration test I think we should use build tags! Please don't use environment variables to trigger the test unless this is needed for configuring the server etc, still I think build tags are the way to go.

Whatever the boss (@berndverst) says :) However, the downside of build tags is that you need to configure your IDE too or the language server won't scan them...

**Review by ItalyPaleAle — APPROVED** [review]

**Review by mukundansundar — APPROVED** [review]

**Comment by 1046102779** [comment]
> IPFS_TEST

the env variable `IPFS_TEST` is a neutral word, if  change to `IPFS_TEST_OFF` is more appropriate?

**Comment by ItalyPaleAle** [comment]
Oh I see now. It is missing a `!` before utils.IsTruthy. Want to open another PR to fix that?

**Comment by 1046102779** [comment]
> Oh I see now. It is missing a `!` before utils.IsTruthy. Want to open another PR to fix that?

hah, i do it.

**Comment by berndverst** [comment]
Once we create certification tests for this component we should move this into a certification test instead! Does someone know how to enable NATS jetstream with a Docker image? Or ideally how to create the Docker-compose file? A lot of configuration seems to be required there.

**Review by berndverst — APPROVED** [review]

**Comment by 1046102779** [comment]
> Once we create certification tests for this component we should move this into a certification test instead! Does someone know how to enable NATS jetstream with a Docker image? Or ideally how to create the Docker-compose file? A lot of configuration seems to be required there.

We may need readme and makefile in this e2e example.

**Pull request merged**